doc: add Customization section and subsections to related sub systems (#50419)
nodejs/node: doc/api/customization.md

# Customization

> Stability: 1.0 - Early development

> Node.js has multiple APIs for customizing behaviour and can easily be invoked

Reviewer: ```suggestion
Node.js has multiple APIs for customizing behavior and can easily be invoked
```

Author: Gah. I think VS Code uses my system language for spell check. (This project has confused the heck outta spell check on iOS)

Reviewer: I use this extension: https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

It allows you to have custom settings by workspace, so you can create a Node workspace and configure `cSpell` within that workspace to use US English.

via [`--import`][]: A `"nodejs-customization"` [entry point][`"exports"`] for a
package loaded via [`--import`][] is automatically run at startup. The
package.json of such a plugin would look something like this:

```json
{
  "name": "example-nodejs-plugin",
  "keywords": [
    "nodejs-plugin"
  ],
  "exports": {
    "nodejs-customization": "./registration.mjs"
  }
}
```

Setting the keyword `nodejs-plugin` is important for users to find the package
(which may be automated). It should also contain the official name(s) for the

> support it provides; for a plugin extending support for typescript would contain
> `typescript` in its keywords; avoid red-herrings such as `typescript` as a
> keyword when the library does not extend support for typescript but is merely
> itself written in typescript. Failing to avoid these red-herrings will cause

Reviewer: ```suggestion
support it provides; a plugin adding Node.js support for TypeScript should contain
`typescript` in its keywords, to avoid matches such as `typescript` as a
keyword when the library does not add support for TypeScript but is merely
itself written in TypeScript. Using nonstandard keywords will cause
```

your package to be listed for things it does not do, aggravating users.

> These steps configure Node.js with a customization plugin:

Reviewer: This section mixes instructions for package authors writing a plugin with instructions for using such a plugin. I feel like those should be separate sections, as end users aren’t going to care about how a plugin should be authored.

Author: I'll add a heading ("Setting up a plugin") between 28 and 30

Reviewer: We should probably put the “Using a plugin” section first, as it applies to more people.

1. Install the plugin via your preferred package manager. For finding the
   plugin, package managers provide often a CLI utility, such as [`npm
   search`][], as well as a website.
1. Once installed, in order to get Node.js to automatically use it, create a
   [`.env`][`--env-file`] file containing an [`--import`][] for your chosen

>    plugin, like `NODE_OPTIONS="--import=example-nodejs-plugin"`.

Reviewer: If we’re using the standardized export that you mentioned above, this would be

```suggestion
   plugin, like `NODE_OPTIONS="--import=example-nodejs-plugin/nodejs-customization"`.
```

Author: OH!! Actually, this is an interesting point, and I think absolves us from a heinous entry-point name: We "control" this `--import`, so if the `--import` itself specifies the entry-point, I think that means it's safe to use whatever we want (so this would be okay):

```json
{
  "name": "example-nodejs-plugin",
  "exports": {
    "register": "whatever.mjs"
  },
  "keywords": ["nodejs-plugin", …]
}
```

```shell
NODE_OPTIONS="--import=example-nodejs-plugin/register"
```

Reviewer: Right, but that’s also why it’s not that important that it be standardized. I think the only benefit of standardization is so that the wizard we eventually create knows what export to use.

Author: > […] the wizard we eventually create knows what export to use.

Yes, exactly.

# Customization

> Stability: 1.0 - Early development

Node.js has multiple APIs for customizing behavior and can easily be invoked via
[`--import`][]. The package.json of such a plugin would look something like
this:

```json
{
  "name": "example-nodejs-plugin",
  "keywords": [
    "nodejs-plugin"
  ],
  "exports": {
    "register": "./registration.mjs"
  }
}
```

Setting the keyword `nodejs-plugin` is important for users to find the
package (which may be automated). It should also contain the official name(s)
for the support it provides; a plugin extending Node.js support for TypeScript
should contain `typescript` in its keywords; a customization package that does
not extend Node.js support for TypeScript should not contain `typescript` (such
as when the library is itself merely written in TypeScript). Incorrect keywords

> will cause a package to be listed for things it does not do, aggravating users.

Reviewer 2: nit: I'm not fan of documenting non-technical patterns in the Node documentation.

Adding the npm keyword is useless from a technical perspective, whether users want to use them or not is more a subjective guideline than a "you must do this for the loader to be compatible with Node.js".

Reviewer: The plan is that it would be a requirement for the wizard we plan to build, so that the wizard can find packages to suggest. But to your point, perhaps it’s better to exclude this until the wizard actually exists, and any steps that are required specifically for wizard compatibility (as opposed to general Node compatibility) could be added later as part of the PR that adds the wizard.

Author: > I'm not fan of documenting non-technical patterns in the Node documentation.

@arcanis are you talking about this part?

> Incorrect keywords will cause a package to be listed for things it does not do, aggravating users.

I'm trying to avoid users complaining (to node) of erroneous results from authors throwing in red-herrings.

---

What is described here is critical for the automated process (in whatever form). To limit changes for authors, these instructions need to be available as soon as possible.

> Adding the npm keyword is useless from a technical perspective, whether users want to use them or not is more a subjective guideline than a "you must do this for the loader to be compatible with Node.js".

I'm confused—this seems to directly contradict what you told me:

> I think uses the npm tags as part of the search, so if all loaders have similar tags that will probably help

So unless what Geoffrey said above (that tags are just keywords lifted from package.json) is incorrect, this is necessary.

The tag needs to be "claimed" immediately, regardless of how it gets there.

## Setting up a plugin

These steps configure Node.js with a customization plugin:

1. Install the plugin via your preferred package manager. For finding the
   plugin, package managers provide often a CLI utility, such as [`npm
   search`][], as well as a website.

> 2. Once installed, in order to get Node.js to use it, you have 2 options:

Reviewer 2: A third option for Yarn users is to add the `NODE_OPTIONS` variable to `.env.yarn`, which gets [automatically injected into all `yarn node` calls](https://yarnpkg.com/configuration/yarnrc#injectEnvironmentFiles).

Author: Hmm, this may be more convenient, but if it works for only a minority of uses, I think it would be better for node to have 1 implementation that works for all.
